x/sync/README.md

```diff
@@ -1,3 +1,124 @@
+If the server had sufficient history to generate a change proof, it contains a change proof that contains 
+the key-value pairs that changed between the requested roots.
+If the server did not have sufficient history to generate a change proof, it contains a range proof that
+contains the key-value pairs that were in the database when the database had the latter root.
+Like range proofs, if a client requests a change proof for range [`requested_start`, `requested_end`] but
+the server can't fit all the key-value pairs in one response,
+it'll send a change proof for [`requested_start`, `proof_end`] where `proof_end` < `requested_end`, 
+as opposed to sending a change proof for [`proof_start`, `requested_end`] where `proof_start` > `requested_start`.
+
+## Algorithm
+
+For each proof it receives, the sync client tracks the root hash of the revision associated with the proof's key-value pairs.
+For example, it will store information that says something like, "I have all of the key-value pairs that
+are in range [`start`, `end`] for the revision with root `root_hash`" for some keys `start` and `end`.
+Note that `root_hash` is the root hash of the revision that the client is trying to sync to, not the 
+root hash of its own (incomplete) database.
+Tracking the revision associated with each downloaded key range, as well as using data in its own 
+(incomplete) database, allows the client to figure out which key ranges are not up to date and need to be synced.
+The hash of the incomplete database on a client is never sent anywhere because it does not represent a root hash of any revision.
+
+When the client is created, it is given the root hash of the revision to sync to.
+When it starts syncing, it requests from a server a range proof for the entire database.
+The server replies with a range proof, which the client verifies.
+If it's valid, the key-value pairs in the proof are written to the database.
+If it's not, the client drops the proof and requests the proof from another server. 
+
+A range proof sent by a server must return a continuous range of the key-value pairs, but may not 
+return the full range that was requested.
+For example, a client might request all the key-value pairs in [`requested_start`, `requested_end`] 
+but only receive those in range [`requested_start`, `proof_end`] where `proof_end` < `requested_end`.
+There might be too many key-value pairs to include in one message, or the server may be too busy to provide any more in its response.
+Unless the database is very small, this means that the range proof the client receives in response to
+ its range proof request for the entire database will not contain all of the key-value pairs in the database.
+
+If a client requests a range proof for range [`requested_start`, `requested_end`] but only receives 
+a range proof for [`requested_start`, `proof_end`] where `proof_end` < `requested_end`
```

Comment:
Our previous discussion might require some amendments here, since `proof_end` probably isn't provided so the comparison isn't possible in most range proof cases.

Reply:
Well, I suppose it depends on what your definition of `proof_end` is. If your definition of `proof_end` is "the greatest key such that the key's value or non-existence is shown by the proof" then all range proofs have `proof_end`. _However_, not all range proofs necessarily have and end proof (as defined by the `RangeProof` type).